kernel/mlx_ethernet/src/work_queue.rs

```diff
@@ -0,0 +1,405 @@
+    /// A multi-part field:
+    /// * `qp_or_sq`: QP/SQ number this WQE is posted to, occupies bits [31:8]
+    /// * `ds`: WQE size in octowords (16-byte units), occupies bits [5:0]
+    pub(crate) ds:                  Volatile<U32<BigEndian>>,
+    /// A multi-part field:
+    /// * `ce`: A value of the type [`CompletionAndEventMode`], occupies bits [3:2]
+    /// * `se`: true if a solicited event, occupies bit 1
+    ce_se:                          Volatile<U32<BigEndian>>,
+    /// general identifier according to WQE opcode/opc_mod
+    ctrl_general_id:                Volatile<U32<BigEndian>>,
+}
+
+const_assert_eq!(core::mem::size_of::<ControlSegment>(), 16);
+
+impl ControlSegment {
+    /// Initialize the fields of the control segment.
+    /// 
+    /// # Arguments
+    /// * `opcode`: the type of command this WQE will complete
+    /// * `wqe_index`: WQEBB number of the first block of this WQE
+    /// * `sqn`: number of the SQ this WQE is posted to
+    fn init(&mut self, opcode: WQEOpcode, wqe_index: u32, sqn: u32) {
+        const WQE_INDEX_SHIFT: u32 = 8;
+        const SQN_SHIFT: u32 = 8;
+        // WQE size in octowords (16-byte units)
+        const WQE_SIZE_IN_OCTWORDS: u32 =  core::mem::size_of::<WorkQueueEntrySend>() as u32 / 16;
+        const CE_SHIFT: u32 = 2;
+
+        self.opcode.write(U32::new((wqe_index << WQE_INDEX_SHIFT)| (opcode as u32)));
+        self.ds.write(U32::new((sqn << SQN_SHIFT) | WQE_SIZE_IN_OCTWORDS));
+        self.ce_se.write(U32::new((CompletionAndEventMode::CQEAlways as u32) << CE_SHIFT));
+    }
+
+    /// Initialize the fields of the control segment to send a packet.
+    /// 
+    /// # Arguments
+    /// * `wqe_index`: WQEBB number of the first block of this WQE
+    /// * `sqn`: number of the SQ this WQE is posted to
+    pub fn send(&mut self, wqe_index: u32, sqn: u32) {
+        self.init(WQEOpcode::Send, wqe_index, sqn)
+    }
+
+    /// Initialize the fields of the control segment to create a NOP.
+    /// With a NOP, only a completion event will be generated. 
+    /// It's a good way to make sure the CQ and WQ are properly initialized.
+    /// 
+    /// # Arguments
+    /// * `wqe_index`: WQEBB number of the first block of this WQE
+    /// * `sqn`: number of the SQ this WQE is posted to
+    pub fn nop(&mut self, wqe_index: u32, sqn: u32) {
+        self.init(WQEOpcode::Nop, wqe_index, sqn)
+    }
+}
+
+/// This segment contains stateless offloading control and inlined Ethernet packet headers
+#[derive(FromBytes, Default)]
+#[repr(C)]
+pub(crate) struct EthSegment {
+    _padding0:              u32,
+    /// Maximum Segment Size
+    mss:                    Volatile<U32<BigEndian>>,
+    flow_table_metadata:    Volatile<U32<BigEndian>>,
+    /// A multi-part field:
+    /// * `inline_header_size`: length of inlined packet headers in bytes, occupies bits [25:16]
+    /// * `inline_headers`: beginning of the inlined packet headers, occupies bits [15:0]
```

Comment:
Should probably break this apart into an actual struct, otherwise it's quite hard to access

Reply:
I prefer keeping fields at the u32 granularity. See my response to the comment above.